parcel/parcel.jl

```diff
@@ -89,68 +94,125 @@ function parcel_model(dY, Y, p, t)
             CMO.a_w_xT(H2SO4_prs, tps, x_sulph, T) - CMO.a_w_ice(tps, T) :
             CMO.a_w_eT(tps, e, T) - CMO.a_w_ice(tps, T)
         J_immersion = CMI_het.ABIFM_J(aerosol, Δa_w)
-        if "Monodisperse" in droplet_size_distribution && "ImmersionFreezing" in ice_nucleation_modes
+        if "Monodisperse" in droplet_size_distribution
             r_l = cbrt(q_liq / N_liq / (4 / 3 * π) / ρ_liq * ρ_air)
-            A_aer = 4 * π * r_l^2
-            dN_act_dt_immersion = max(FT(0), J_immersion * N_liq * A_aer)
-            dqi_dt_new_immers = dN_act_dt_immersion * 4 / 3 * π * r_l^3 * ρ_ice / ρ_air
+        elseif "Gamma" in droplet_size_distribution
+            λ = cbrt(32 * π * N_liq / q_liq * ρ_liq / ρ_air)
+            #A = N_liq* λ^2
+            r_l = 2 / λ
         end
-        if "Gamma" in droplet_size_distribution && "ImmersionFreezing" in ice_nucleation_modes
+        A_aer = 4 * π * r_l^2
+        dN_act_dt_immersion = max(FT(0), J_immersion * N_liq * A_aer)
+        dqi_dt_new_immers = dN_act_dt_immersion * 4 / 3 * π * r_l^3 * ρ_ice / ρ_air
+    end
+
+    dN_act_dt_hom = FT(0)
+    dqi_dt_new_hom = FT(0)
+    if "HomogeneousFreezing" in ice_nucleation_modes
+        a_w_ice = CMO.a_w_ice(tps, T)
+        Δa_w = CMO.a_w_eT(tps, e, T) - a_w_ice
+        if "Monodisperse" in droplet_size_distribution
+            J_hom = CMI_hom.homogeneous_J(ip.homogeneous, Δa_w)
+            r_l = cbrt(q_liq / N_liq / (4 / 3 * π) / ρ_liq * ρ_air)
+            V_l = 4 /3 * π * r_l^3
+            dN_act_dt_hom = max(FT(0), J_hom * N_liq * V_l)
+            dqi_dt_new_hom = dN_act_dt_hom * 4 / 3 * π * r_l^3 * ρ_ice / ρ_air
+        elseif "Gamma" in droplet_size_distribution
+            J_hom = CMI_hom.homogeneous_J(ip.homogeneous, Δa_w)
             λ = cbrt(32 * π * N_liq / q_liq * ρ_liq / ρ_air)
             #A = N_liq* λ^2
             r_l = 2 / λ
-            A_aer = 4 * π * r_l^2
-            dN_act_dt_immersion = max(FT(0), J_immersion * N_liq * A_aer)
-            dqi_dt_new_immers = dN_act_dt_immersion * 4 / 3 * π * r_l^3 * ρ_ice / ρ_air
+            V_l = 4 / 3 * π * r_l^3
+            dN_act_dt_hom = max(FT(0), J_hom * N_aer * V_l)
+            dqi_dt_new_hom = dN_act_dt_hom * 4 / 3 * π * r_l^3 * ρ_ice / ρ_air
+        elseif "Lognormal" in droplet_size_distribution
+            J_hom = CMI_hom.homogeneous_J(ip.homogeneous, Δa_w)
+            dN_act_dt_hom = max(FT(0), J_hom * N_liq * 4 / 3 * π * exp((6*log(R_mode_liq) + 9 * σ^2)/(2)))
+            r_l = R_mode_liq * exp(σ)
+            dqi_dt_new_hom = dN_act_dt_hom * 4 / 3 * π * r_l^3 * ρ_ice / ρ_air
+        elseif "Jensen_Example" in droplet_size_distribution
+            if Δa_w < ip.homogeneous.Δa_w_min
+                Δa_w = ip.homogeneous.Δa_w_min
+            elseif Δa_w > ip.homogeneous.Δa_w_max
```

Comment:
It might be better for GPUs to do it like this: 
` Δa_w = ifelse(Δa_w < ip.homogeneous.Δa_w_min, ip.homogeneous.Δa_w_min, ifelse(Δa_w > ip.homogeneous.Δa_w_max, ip.homogeneous.Δa_w_max, Δa_w))`